go/proxyd/cache.go

```diff
@@ -0,0 +1,275 @@
+package proxyd
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/go-redis/redis/v8"
+	"github.com/golang/snappy"
+	lru "github.com/hashicorp/golang-lru"
+)
+
+type Cache interface {
+	Get(ctx context.Context, key string) (string, error)
+	Put(ctx context.Context, key string, value string) error
+}
+
+// assuming an average RPCRes size of 3 KB
+const (
+	memoryCacheLimit      = 4096
+	numBlockConfirmations = 50
+)
+
+var (
+	supportedRPCMethods = map[string]bool{
+		"eth_chainId":          true,
+		"net_version":          true,
+		"eth_getBlockByNumber": true,
+		"eth_getBlockRange":    true,
+	}
+	supportedBlockRPCMethods = map[string]bool{
+		"eth_getBlockByNumber": true,
+		"eth_getBlockRange":    true,
+	}
+)
+
+var (
+	errInvalidBlockByNumberParams = errors.New("invalid eth_getBlockByNumber params")
+	errUnavailableBlockNumSyncer  = errors.New("getLatestBlockFn not set for required RPC")
+)
+
+type cache struct {
+	lru *lru.Cache
+}
+
+func newMemoryCache() *cache {
+	rep, _ := lru.New(memoryCacheLimit)
+	return &cache{rep}
+}
+
+func (c *cache) Get(ctx context.Context, key string) (string, error) {
+	if val, ok := c.lru.Get(key); ok {
+		return val.(string), nil
+	}
+	return "", nil
+}
+
+func (c *cache) Put(ctx context.Context, key string, value string) error {
+	c.lru.Add(key, value)
+	return nil
+}
+
+type redisCache struct {
+	rdb *redis.Client
+}
+
+func newRedisCache(url string) (*redisCache, error) {
+	opts, err := redis.ParseURL(url)
+	if err != nil {
+		return nil, err
+	}
+	rdb := redis.NewClient(opts)
+	if err := rdb.Ping(context.Background()).Err(); err != nil {
+		return nil, wrapErr(err, "error connecting to redis")
+	}
+	return &redisCache{rdb}, nil
+}
+
+func (c *redisCache) Get(ctx context.Context, key string) (string, error) {
+	val, err := c.rdb.Get(ctx, key).Result()
+	if err == redis.Nil {
+		return "", nil
+	} else if err != nil {
+		return "", err
+	}
+	return val, nil
+}
+
+func (c *redisCache) Put(ctx context.Context, key string, value string) error {
+	err := c.rdb.Set(ctx, key, value, 0).Err()
+	return err
+}
+
+type GetLatestBlockNumFn func(ctx context.Context) (uint64, error)
+
+type RPCCache struct {
+	cache               Cache
+	getLatestBlockNumFn GetLatestBlockNumFn
+}
+
+func newRPCCache(cache Cache, getLatestBlockNumFn GetLatestBlockNumFn) *RPCCache {
+	return &RPCCache{cache: cache, getLatestBlockNumFn: getLatestBlockNumFn}
+}
+
+func (c *RPCCache) GetRPC(ctx context.Context, req *RPCReq) (*RPCRes, error) {
+	if !c.isCacheable(req) {
+		return nil, nil
+	}
+
+	key := mustMarshalJSON(req)
```

Comment:
I wouldn't use the marshalled JSON as the cache key, since it contains an ID that will change on each request. Instead, I'd define a per-method cacheKey. For example, the cacheKey for `eth_getBlockByNumber` might be `method:eth_getBlockByNumber:123` (assuming that the RPC request is a valid one).

Reply:
good catch! Similarly, I'll have to cache RPC responses appropriately.